### PR TITLE
Introduce a new SSA-based intermediate format for the compiler

### PR DESCRIPTION
**Pull request by bjorng**
This pull request introduces a Static Single Assignment (SSA) format as a new intermediate format in the compiler. The code generator as well as several optimization passes has been completely rewritten.
    
It is easier to write many kind optimizations for new SSA-based intermediate format than for the BEAM assembly language. We plan to introduce significantly improved optimizations of binary matching later this year.

For more details about the new compiler passes, see the individual commit messagges in this branch (especially "Introduce a new SSA-based intermediate format").

We hope to merge this pull request in one or two weeks. We'll try to fix any major bugs (compiler crashes or incorrectly generated code) before merging.

### CODE REVIEW

File: lib/compiler/src/beam_z.erl [hunk 1]
```diff
@@ -110,6 +110,8 @@ undo_rename({test,has_map_fields,Fail,[Src|List]}) ->
     {test,has_map_fields,Fail,Src,{list,List}};
 undo_rename({get_map_elements,Fail,Src,{list,List}}) ->
     {get_map_elements,Fail,Src,{list,List}};
+undo_rename({test,is_eq_exact,Fail,[Src,nil]}) ->
+    {test,is_nil,Fail,[Src]};
```
[the remark this answers is not in the document]
Comment (bjorng) [reply lib/compiler/src/beam_z.erl]: > Is it possible the `nil` value will be the first argument?

No, that should not be possible.

There is this clause in `beam_utils:bif_to_test/3` that places any literal argument in the second position:

```
bif_to_test('=:=', [C,A], Fail) when ?is_const(C) ->
    {test,is_eq_exact,Fail,[A,C]};
```

Here is the definition of `is_const`:

```
-define(is_const(Val), (Val =:= nil orelse
                        element(1, Val) =:= integer orelse
                        element(1, Val) =:= float orelse
                        element(1, Val) =:= atom orelse
                        element(1, Val) =:= literal)).
```

[the remark this answers is not in the document]
Comment (michalmuskala) [reply lib/compiler/src/beam_z.erl]: Awesome. Sorry for the confusion.

**Comment by josevalim** [comment]
Hi @bjorng! This looks very interesting, happy to see it merged! Are there any existing passes that you suspect that it could be cleaner (i.e. less lines of code) or more efficient (i.e. applies in more cases) if moved to SSA? If so, I would love to explore it, as a way to get more familiar with how SSA works. Thanks!

**Comment by bjorng** [comment]
Here are some passes that I think could benefit from SSA:

`sys_core_dsetel`. Could be an easy way to get started with SSA. You could also implement #1890.

The sub pass `v3_kernel:guard_opt/2`.

`beam_trim`. This probably has to be implemented as a sub pass of `ssa_codegen` or `ssa_pre_codegen`.

The minor optimizations in `beam_flatten`. They probably have to be done in `beam_ssa_codegen`.

Optimization of `#b_switch{}` branches. If two branches jumps to code blocks that do the same thing, let both branches jump to the same block. `beam_jump` does this kind of optimization, but doing it earlier in the SSA format could speed up compilation of function with many clauses.

Getting rid of the call to `beam_utils:is_killed/3` in `beam_jump` (in some way to be determined). We want to remove `beam_utils:is_killed()`.

A pass I am not sure about is `beam_bs`. We could possibly gain some compilation speed by rewriting it in SSA. Could also be a good project for getting started with SSA.

I am currently working on a replacement for `beam_dead`. John is working on a replacement for `beam_bsm`.

**Pull request merged**